capstone/src/test/java/com/google/sps/servlets/JoinGroupServletTest.java

```diff
@@ -0,0 +1,203 @@
+
+  private static final String USER_EMAIL = "test@test.com";
+  private static final String USER_ID = "test";
+  private static final String GROUP_ID = "1";
+  private static final String GROUP_NAME = "The Bachelors";
+  private static final String FIRST_NAME = "Lucy";
+  private static final String LAST_NAME = "Qu";
+
+  private final LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(
+              new LocalDatastoreServiceTestConfig()
+                  .setDefaultHighRepJobPolicyUnappliedJobPercentage(0),
+              new LocalUserServiceTestConfig())
+          .setEnvEmail(USER_EMAIL)
+          .setEnvIsLoggedIn(true)
+          .setEnvAuthDomain("gmail.com")
+          .setEnvAttributes(
+              new HashMap(
+                  ImmutableMap.of(
+                      "com.google.appengine.api.users.UserService.user_id_key", USER_ID)));
+
+  @Mock private HttpServletRequest mockRequest;
+  @Mock private HttpServletResponse mockResponse;
+  private JoinGroupServlet joinGroupServlet;
+  private StringWriter responseWriter;
+  private DatastoreService datastore;
+
+  @Before
+  public void setUp() throws Exception {
+    MockitoAnnotations.initMocks(this);
+    helper.setUp();
+    datastore = DatastoreServiceFactory.getDatastoreService();
+    joinGroupServlet = new JoinGroupServlet();
+
+    // Set up a fake HTTP response.
+    responseWriter = new StringWriter();
+    when(mockResponse.getWriter()).thenReturn(new PrintWriter(responseWriter));
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+    responseWriter = null;
+    datastore = null;
+    joinGroupServlet = null;
+  }
+
+  private Entity createGroup(String userId, String groupName, List<String> members) {
+    Entity groupEntity = new Entity("Group");
+    groupEntity.setProperty("memberIds", members);
+    groupEntity.setProperty("challenges", null);
+    groupEntity.setProperty("posts", null);
+    groupEntity.setProperty("options", new ArrayList<Long>());
+    groupEntity.setProperty("groupName", groupName);
+    groupEntity.setProperty("headerImg", "");
+    return groupEntity;
+  }
+
+  private Entity createUser(String userId, String firstName, String lastName) {
+    Entity userEntity = new Entity("User", userId);
+    userEntity.setProperty("userId", userId);
+    userEntity.setProperty("firstName", firstName);
+    userEntity.setProperty("lastName", lastName);
+    userEntity.setProperty("email", "");
+    userEntity.setProperty("phoneNumber", "");
+    userEntity.setProperty("profilePic", "");
+    userEntity.setProperty("badges", null);
+    userEntity.setProperty("groups", null);
+    userEntity.setProperty("interests", null);
+    return userEntity;
+  }
+
+  @Test
+  public void doGet_userIsMember() throws IOException {
+    List<String> members = new ArrayList<String>();
+    members.add(USER_ID);
+    Entity groupEntity = this.createGroup(USER_ID, GROUP_NAME, members);
+    datastore.put(groupEntity);
```

Comment:
I think this might pass anyway because how you're doing the check in JoinGroupServlet, but shouldn't you also add the groupId to the user's list of groups?

Reply:
Wait, I see that you tested for that later on, so maybe it's not necessary.

Reply:
Yeah, in my getIsMember() method on JoinGroupServlet, I only check whether the group's list of members contains the member. I do update the user's list of groups, but that's not part of my check for whether the user is a member, so that is why I have two separate tests